votetimer.py

```diff
@@ -0,0 +1,318 @@
+        self.queue_next_time(town_id, end_time, now)
+        self.town_ticker.start_ticking()
+        return ret
+
+    async def remove_town(self, town_id):
+        had_town = False
+        if town_id in self.town_map:
+            had_town = True
+            self.town_map.pop(town_id)
+            self.town_storage.remove_town(town_id)
+
+        if not self.town_storage.has_towns_ticking():
+            self.town_ticker.stop_ticking()
+        
+        if had_town:
+            town_info = self.town_info_provider.get_town_info(town_id)
+            message = f'{town_info.villager_role.mention} - Vote countdown stopped!'
+            await self.message_broadcaster.send_message(town_info, message)
+
+    async def tick(self):
+        finished = self.town_storage.tick_and_return_finished_towns()
+
+        for town_id in finished:
+            await self.advance_town(town_id, self.datetime_provider.now())
+        
+        if not self.town_storage.has_towns_ticking():
+            self.town_ticker.stop_ticking()
+
+    async def advance_town(self, town_id, now):
+        if (town_id in self.town_map):
+            end_time = self.town_map[town_id]
+            if (end_time < now):
+                self.town_map.pop(town_id)
+                await self.vote_handler.perform_vote(town_id)
+            else:
+                await self.send_message(town_id, end_time, now)
+                self.queue_next_time(town_id, end_time, now)
+
+    async def send_message(self, town_id, end_time, now):
+        town_info = self.town_info_provider.get_town_info(town_id)
+        message = self.construct_message(town_info, end_time, now)
+        return await self.message_broadcaster.send_message(town_info, message)
+
+    def queue_next_time(self, town_id, end_time, now):
+        next_time = end_time
+        delta = (end_time - now).total_seconds()
+        if delta >= 0:
+            advance_times = [300, 60, 15, 0]
+            for x in advance_times:
+                if delta > x:
+                    next_time = end_time - datetime.timedelta(seconds=x)
+                    break
+
+        self.town_storage.add_town(town_id, next_time)
+
+    def construct_message(self, town_info, end_time, now):
+        delta_seconds = (end_time - now).total_seconds()
+
+        rounded_seconds = round(delta_seconds/5)*5
+
+        message = f'{town_info.villager_role.mention} - '
+
+        if rounded_seconds > 0:
+            minutes = math.floor(rounded_seconds / 60)
+            seconds = rounded_seconds % 60
+
+            if minutes > 0:
+                message = message + f'{minutes} minute'
+                if minutes > 1:
+                    message = message + 's'
+                if seconds > 0:
+                    message = message + ', '
+
+            if seconds > 0:
+                message = message + f'{seconds} seconds'
+
+            message = message + ' remaining!'
+        else:
+            message = message + 'Time to vote!'
```

Comment:
This never got hit in my tests, but it is possible if there's a big delay and some earlier messages got skipped, so I left it in. I could remove it.